api/docs/scatter_gather_emulation.dox

```diff
@@ -71,6 +74,126 @@ register `k1`. Elements may be scattered in any order. When an element is stored
 its mask is cleared. If some store faults, all elements to its right (closer to
 LSB) will be complete.
 
+## AArch64
+
+The AArch64 SVE (Scalable Vector Extension) introduced the first AArch64 scatter and
+gather instructions.
+
+SVE has two scatter and gather addressing modes (scalar+vector and vector+immediate)
+and two predicated contiguous load and store addressing modes (scalar+scalar and
+scalar+immediate). The SVE2 instruction set extension adds a third scatter/gather
+addressing mode: vector+scalar.
+
+The predicated contiguous instructions do not use vector based addressing but they do
+have other similarities with the scatter and gather instructions which mean DynamoRIO
+handles them in a similar way.
+
+All scatter/gather/predicated contiguous instructions access elements conditionally based
+on the mask value of a governing predicate register. Loads are always zeroing: inactive
+elements are set to 0 in the destination register.
+
+### Scalar+vector
+
+```
+ld1b (%x0,%z0.s,uxtw)[1byte] %p0/z -> %z1.s
+```
+
+Above is a scalar+vector gather load that reads 8-bit values which are zero-extended and
+written to 32-bit elements of the `z1` vector register. Addresses are calculated by
+adding the base address from the scalar `x0` register to the corresponding 32-bit element
+in the vector index register `z0`. Elements that are not active in the mask contained by
+the governing predicate register `p0` are not loaded and the corresponding element in the
+destination register `z1` is set to `0`.
+
+```
+st1h   %z17.d %p5 -> (%x19,%z20.d)[2byte]
+```
+
+Above is a scalar+vector scatter store that writes the lowest 16-bits of the 64-bit
+elements of the `z17` vector register. Addresses are calculated by adding the base
+address from the scalar `x19` register to the corresponding 64-bit element in the vector
+index register `z20`. Elements that are not active in the mask contained by the governing
+predicate register `p5` are not stored.
+
+### Vector+immediate
+
+```
+ld1sw  +0x18(%z8.d)[4byte] %p2/z -> %z6.d
+```
+Above is a vector+immediate gather load that reads 32-bit values which are sign-extended
+and written to 64-bit elements of the `z6` vector register. Addresses are calculated by
+adding the immedate value 0x18 to the corresponding 64-bit element in the vector base
+register `z8`. Elements that are not active in the mask contained by governing predicate
+register `p2` are zeroed.
+
+### Vector+scalar
+
+```
+stnt1d %z27.d %p7 -> (%z25.d,%x29)[8byte]
```

Comment:
Can the vector reg suffix be different than the opcode? E.g., can we use .h here? which would mean storing 16-bit vector elements to 64-bit memory elements?

Reply:
Yes it can, but not in the way you suggest. Which combinations are valid depends on the instruction but in general the element size must always be >= to the opcode value size.
So you can't have a store that stores 16-bit vector elements to 64-bit memory locations.
`stnt1d %z27.h %p7 -> (%z25.d,%x29)[8byte] // invalid`
But you can have a store that stores the low 16-bits of 64-bit elements to 16-bit memory locations:
`stnt1h %z27.d %p7 -> (%z25.d,%x29)[8byte] // valid`


For example scalar+scalar `st1h`:
`ST1H { <Zt>.<T> }, <Pg>, [<Xn|SP>, <Xm>, LSL #1]`

allows `.h`, `.s`, or `.d` element sizes. For the `.s` and `.d` variants it stores the lowest halfword (16-bits) of each element.

Reply:
> So you can't have a store that stores 16-bit vector elements to 64-bit memory locations.

Okay. I may have guessed that the remaining part of the 64-bit memory location is zero-extended or something. Thanks for clarifying.